---
title: "ProjetII"
author: "Sven Sp√∂rri, Thomas Castiglione"
date: "07/12/2020"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# R environment

We use the package ggplot2

```{r}
library(ggplot2)
```

# 2. Data exploration

We load the data genotypes.txt, variants_info_openSNP.txt, phenotypes.txt and covariates.txt, and change the data format to be more suitable for our work.

```{r}
genotypes<-read.table('ressources/genotypes.txt',header=T)
phenotypes<-read.table('ressources/phenotypes.txt', header=T)
variants.info<-read.table('ressources/variants_info_openSNP.txt',header=T)
covariates<-read.table('ressources/covariates.txt',header=T)

```

1. linear regression.
```{r}
covariates.pheno<-cbind(covariates,phenotypes$HEIGHT)
rownames(covariates.pheno)<-covariates$CHALLENGE_ID
covariates.pheno<-covariates.pheno[,-1]
colnames(covariates.pheno)<-c('SEX','PLATFORM','HEIGHT')

lin.reg.cov<-lm(HEIGHT~., data=covariates.pheno)
summary(lin.reg.cov)

```
The multiple R-squared value of 0.472 shows us that the covariates does impact the phenotype, but it is does not fully determines it.

2. Plotting the phenotype distribution
```{r}
ggplot(phenotypes)+geom_density(aes(HEIGHT),size=1)+theme_bw()
```
```{r}
covariates.pheno.M<-covariates.pheno[apply(covariates.pheno,1,function(x) x[1]=='M'),]
covariates.pheno.F<-covariates.pheno[apply(covariates.pheno,1,function(x) x[1]=='F'),]

ggplot()+geom_density(data=covariates.pheno,aes(HEIGHT,color='male+female'),size=1.1)+
  geom_density(data=covariates.pheno.M,aes(HEIGHT,color='male'),size=0.8)+
  geom_density(data=covariates.pheno.F,aes(HEIGHT,color='female'),size=0.8)+ theme_bw()

```
# Genome-Wide Association Study

1.
running GWAS without covariates
```{r}
gwas<-function(SNP,phenos=phenotypes$HEIGHT){
  df<-data.frame(cbind(SNP,phenos))
  colnames(df)<-c('SNP','HEIGHT')
  fit<-lm(HEIGHT~.,data=df)
  summary.coefs<-summary(fit)$coefficient
  data.frame(coef=summary.coefs[2,1],intercept=summary.coefs[1,1],pval=summary.coefs[2,4])
}
gwas.with.cov<-function(SNP,cov=covariates.pheno){
  df<-data.frame(cbind(SNP,cov))
  colnames(df)<-c('SNP',colnames(covariates.pheno))
  fit<-lm(HEIGHT~.,data=df)
  summary.coefs<-summary(fit)$coefficients
  data.frame(coef=summary.coefs[2,1],intercept=summary.coefs[1,1],pval=summary.coefs[2,4])
}
```


```{r}
gwas.res<-data.frame(coef=NULL,pval=NULL)
for(i in 1:ncol(genotypes)){
  gwas.res<-rbind(gwas.res,gwas(genotypes[,i]))
}
rownames(gwas.res)<-colnames(genotypes)
gwas.res<-cbind(gwas.res, variants.info$CHR)
```

running GWAS with covariates
```{r}
gwas.res.with.cov<-data.frame(coef=NULL,pval=NULL)
for(i in 1:ncol(genotypes)){
  gwas.res.with.cov<-rbind(gwas.res.with.cov,gwas.with.cov(genotypes[,i]))
}
rownames(gwas.res.with.cov)<-colnames(genotypes)
gwas.res.with.cov<-cbind(gwas.res.with.cov,variants.info$CHR)
```

2.
Manhattan plots.
```{r}
manhattan.plot<-function(gwas.out,title){
colnames(gwas.out)<-c('coef','intercept','pval','CHR')
bonferroni_thresh<-0.05/nrow(gwas.out)
ggplot(gwas.out,aes(x=1:nrow(gwas.out),y=-log10(pval),color=CHR%%2,size=-log10(pval)))+
  geom_point(shape=20)+
  geom_hline(yintercept = -log10(bonferroni_thresh),linetype='dotdash')+
  scale_size_continuous(range = c(1,4))+
  scale_y_continuous(limits=c(0,10))+
  theme(legend.position = 'none')+
  annotate(geom='text', x=600,y=5.6, label='Bonferroni threshold',size=3.5)+
  labs(x='SNP position',y='-log10(P-value)')+
  ggtitle(title)
}
```

```{r}
manhattan.plot(gwas.res, 'GWAS Visualisation without covariates')
```


```{r}
manhattan.plot(gwas.res.with.cov, 'GWAS Visualitation with covariates')
```

#  Meta-analysis with the GIANT Study


```{r}
# read data from giant study
giant.summary <- read.table('ressources/summary_GWAS_giant.txt',header=T)
print(paste0('# of SNPs not available in GIANT dataframe: ',nrow(gwas.res.with.cov) - nrow(giant.summary)))
# add rownames as column to allow merge on column SNP
gwas.res.with.cov.new <- cbind(data.frame(SNP=rownames(gwas.res.with.cov)),gwas.res.with.cov)
# merge giant study data with our data
merged <- merge(giant.summary,gwas.res.with.cov.new,by='SNP')
# linear fit between coefficients from giant study and our analysis
coefs <- summary(lm(merged$coef~merged$BETA))$coefficients
#plot
ggplot(merged,aes(x=BETA,y=coef)) +
  geom_point() +
  geom_abline(intercept=coefs[1,1],slope=coefs[2,1]) +
  xlab(paste0("\u03B2", '(GIANT study)')) + ylab(paste0("\u03B2", '(our analysis)'))
print('Correlation coefficient: ')
cor.test(merged$coef,merged$BETA)
```

# Genomic prediction

```{r}
genotypes.test <- read.table('ressources/genotypes_test.txt',header=T)
res <- as.matrix(genotypes.test[,giant.summary$SNP]) %*% as.matrix(gwas.res.with.cov[giant.summary$SNP,]$coef)

hist(res)



head(giant.summary)

length(giant.summary$BETA)
head(giant.summary)

res.giant <- as.matrix(genotypes.test[giant.summary$SNP]) %*% as.matrix(giant.summary$BETA)
hist(res.giant)

fit <- lm(HEIGHT~.,data=data.frame(HEIGHT=phenotypes$HEIGHT,genotypes))
res <- predict(fit,data=data.frame(genotypes.test))
summary(fit)['r.squared']
```


```{r}

```

